effects: taint `:nothrow` effect on unknown `:static_parameter`

aviatesk opened the pull request:
With this commit, we taint `:nothrow` effect property correctly on access to unknown `:static_parameter`, e.g.:
```julia
unknown_sparam_throw(::Union{Nothing, Type{T}}) where T = (T; nothing)
@test Core.Compiler.is_nothrow(Base.infer_effects(unknown_sparam_throw, ((Type{Int},))))
@test !Core.Compiler.is_nothrow(Base.infer_effects(unknown_sparam_throw, ((Nothing,))))
```

This commit implements a very conservative analysis, and thus there is a room for improvement still, e.g.:
```julia
unknown_sparam_nothrow(x::Ref{T}) where {T} = (T; nothing)
@test_broken Core.Compiler.is_nothrow(Base.infer_effects(unknown_sparam_nothrow, (Ref,)))
```

fix #46771

aviatesk committed:
effects: taint `:nothrow` effect on unknown `:static_parameter` (conservatively)

With this commit, we taint `:nothrow` effect property correctly on
access to unknown `:static_parameter`, e.g.:
```julia
unknown_sparam_throw(::Union{Nothing, Type{T}}) where T = (T; nothing)
@test Core.Compiler.is_nothrow(Base.infer_effects(unknown_sparam_throw, ((Type{Int},))))
@test !Core.Compiler.is_nothrow(Base.infer_effects(unknown_sparam_throw, ((Nothing,))))
```

This commit implements a very conservative analysis, and thus there is a
room for improvement still, e.g.:
```julia
unknown_sparam_nothrow(x::Ref{T}) where {T} = (T; nothing)
@test_broken Core.Compiler.is_nothrow(Base.infer_effects(unknown_sparam_nothrow, (Ref,)))
```

aviatesk commented:
@Keno is it okay for now to merge this as is, and try to improve the accuracy of this `nothrow` analysis in a follow up PR (it seems a bit tricky than I expected though)?

Keno commented:
I would prefer to try to get at least some precision in at the same time, because I suspect we're relying on this bug for performance in a number of applications

staticfloat commented:
@aviatesk I rebased this branch, so that we can push forward [Keno's PR](https://github.com/JuliaLang/julia/pull/46820) more easily.

vchuravy commented:
Will this need to be backported to 1.9 since it's a bugfix to the effect system?

Keno committed:
inference: improve `:nothrow` modeling for `:static_parameter` (#46820)

aviatesk commented:
> Will this need to be backported to 1.9 since it's a bugfix to the effect system?

Yes, I think we should do it.

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/46791#issuecomment-1420908340) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/52242a5_vs_2c619b7/report.md).

oscardssmith commented:
given that there aren't any regressions here, good to merge?

aviatesk commented:
There is one test failure on this PR, so don't merge now.

staticfloat committed:
Fix test with free type params

Keno committed:
Test: Ignore ::Type{T} in detect_unbounded

These are only technically unbounded because of the existence of
ill-formed types. However, this function is supposed to be an API
sanity check and ordinary users should never have ill-formed types,
so for the purpose we want here, allow unboundedness in Type{T}.

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/46791#issuecomment-1422054540) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/c46b5cc_vs_3a92d38/report.md).
